### DISCUSSION TITLE
How can I sandbox a binary that executes bwrap?

### DISCUSSION THREAD

**Discussion by amano-kenji** [netblue30/firejail#5220] [Q&A]
If I try to execute https://github.com/johnfactotum/foliate inside firejail sandbox, I get this error.

> ** (com.github.johnfactotum.Foliate:9): ERROR **: 06:52:25.348: Failed to start dbus proxy: Failed to spawn child process “/usr/bin/bwrap” (Permission denied)

foliate internally uses webkit, python, and gjs.

Perhaps, I should use apparmor for this?

**Comment by rusty-snake** [comment]
Please include more information (namely software versions, distribution, is bwrap suid, ...).

General

```mermaid
flowchart TB

A{Is bwrap install as suid?}
A -->|Yes| B(Allow suid.<br>This means you must <i>ignore</i>/remove<ul><li>noroot</li><li>nonewprivs</li><li>seccomp</li><li>protocol</li><li>caps.drop all</li><li>...</li></ul>)
A -->|No| C(Allow other things bwrap is relaying on such as mounting.)
```

> Perhaps, I should use apparmor for this?

AppArmor <sub>or SELinux</sub> is easier give that even if bwrap can run there is still some incompatibility if `--unshare-pid --proc /proc` is used (#3647).

**Reply by amano-kenji** [reply]
bubblewrap-0.6.1 on Gentoo Linux.
It was built with SUID.

**Reply by amano-kenji** [reply]
foliate depends on webkit-gtk for rendering epub HTML files.
It turns out that if I build webkit-gtk without seccomp support, webkit-gtk doesn't depend on bubblewrap.

**Reply by rusty-snake** [reply]
The Gentoo solution:

> I just need to change the USE flags.

:rofl: :rofl:

edit: I should make a meme out of it :smile: .

**Reply by rusty-snake** [reply]
>  It turns out that if I build webkit-gtk without seccomp support, webkit-gtk doesn't depend on bubblewrap.

As this weaks the webkit sandbox for all your program that use webkit-gtk you should be careful doing so. As an alternative you can disable the webkit sandbox for foliate. <sub><sup><code>WEBKIT_FORCE_SANDBOX=0</code></sup></sub>

**Reply by amano-kenji** [reply]
Where did you find WEBKIT_FORCE_SANDBOX?

**Reply by rusty-snake** [reply]
It's nowhere documented AFAIK. Because you should not use it unless you understand the consequences.

edit: Actually it is document for value `1`. https://trac.webkit.org/wiki/EnvironmentVariables
edit: And with value `0` for debugging. https://trac.webkit.org/wiki/WebKitGTK/Debugging

**Reply by amano-kenji** [reply]
https://bugs.webkit.org/show_bug.cgi?id=220117 talks about renaming WEBKIT_FORCE_SANDBOX to another name that feels scarier.

**Reply by rusty-snake** [reply]
Because it's dangerous if you don't understand the consequences. That why I don't really push this knowledge (I could have written it in my first answer). See https://github.com/netblue30/firejail/commit/a6084bb054709086ba5d156e5b5b2dafd7323774 for some summary of my opinion.

**Reply by amano-kenji** [reply]
`env WEBKIT_FORCE_SANDBOX=0` inside a firejail sandbox is pretty safe.

**Reply by rusty-snake** [reply]
Not really, if you know how unsafe a firejail sandbox can be.

Generally you should never disable a internal sandbox (firefox, chromium, webkit-gtk, nautilus, eog, ...) unless
1. you understand
2. you have to
3. you use a tight external sandbox 
4. only trusted input is processed

**Reply by amano-kenji** [reply]
com.github.johnfactotum.Foliate.profile is broken without `env WEBKIT_FORCE_SANDBOX=0`. Do you have any better idea?

**Reply by rusty-snake** [reply]
Don't use a external sandbox or use apparmor/selinux or use the flatpak. Or fix bwrap in firejail.

Ok non of the is really helpful (depending).

We should remove it from firecfg.config (at least until bwrap works in firejail).

**Reply by amano-kenji** [reply]
I can learn apparmor a little bit more and become good at creating apparmor profiles. Can SUID bwrap work inside firejail?

It seems bwrap should detect whether it's inside an SUID sandbox.

**Reply by rusty-snake** [reply]
> Can SUID bwrap work inside firejail?

chrome-sandbox (suid) can work inside firejail.

**Reply by amano-kenji** [reply]
Here's my foliate.local

```
env WEBKIT_FORCE_SANDBOX=0
dbus-user filter
dbus-user.own com.github.johnfactotum.Foliate
dbus-user.talk org.freedesktop.portal.Desktop
dbus-user.talk org.freedesktop.PackageKit
dbus-user.talk ca.desrt.dconf
dbus-user.talk org.a11y.*
dbus-system none
```

Because foliate profile has `private-tmp` and `net none`, it doesn't have access to dbus user session by default.

**Reply by amano-kenji** [reply]
With the following lines in foliate.local
```
ignore noroot
ignore nogroups
ignore nonewprivs
ignore seccomp
ignore protocol
ignore caps.drop
private-bin bwrap
noblacklist ${PATH}/bwrap
```
```
$ firejail --profile=/etc/firejail/com.github.johnfactotum.Foliate.profile /usr/bin/com.github.johnfactotum.Foliate
...
bwrap: setting up uid map: Permission denied
```
